### PR TITLE
Implements Datadog as output plugin

### PR DESCRIPTION
**Pull request by clamoriniere**
This PR adds Datadog output plugin.

configuration example:
```
[OUTPUT]
    Name        datadog
    Match       *
    Host        http-intake.logs.datadoghq.com
    TLS         on
    apikey      <my-datadog-api-key>
    dd_service  my-app-service
    dd_source   my-app
    dd_tags     team:logs,foo:bar
```

### CODE REVIEW

File: plugins/out_datadog/datadog.c [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    if (ret == -1) {
+        FLB_OUTPUT_RETURN(FLB_ERROR);
+    }
+
+    /* Create HTTP client context */
+    client = flb_http_client(upstream_conn, FLB_HTTP_POST, ctx->uri,
+                             payload_buf, payload_size,
+                             ctx->host, ctx->port,
+                             NULL, 0);
+
```
Comment (edsiper) [inline plugins/out_datadog/datadog.c]: always check if client is not null, if so, release resources allocated.

File: plugins/out_datadog/datadog.c [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    int64_t timestamp;
+    msgpack_object *obj;
+    msgpack_object map;
+    msgpack_object root;
+    msgpack_object k;
+    msgpack_object v;
+    /* output buffer */
+    flb_sds_t out_buf;
+
+    /* Count number of records */
```
Comment (edsiper) [inline plugins/out_datadog/datadog.c]: to count the number of items, please use the new flb_mp_count() function, defined here:

https://github.com/fluent/fluent-bit/blob/master/src/flb_mp.c#L24

File: plugins/out_datadog/datadog.c [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        msgpack_pack_str_body(&mp_pck,
+                              ctx->json_date_key,
+                              flb_sds_len(ctx->json_date_key));
+        msgpack_pack_int64(&mp_pck, timestamp);
+
+        /* include_tag_key */
+        if (ctx->include_tag_key == FLB_TRUE) {
+            dd_msgpack_pack_key_value_str(&mp_pck,
+                                          ctx->tag_key, flb_sds_len(ctx->tag_key),
+                                          tag, sizeof(tag_len)-1);
```
Comment (edsiper) [inline plugins/out_datadog/datadog.c]: sizeof(tag_len) will become the data type size, tag_len should be fine

**Comment by edsiper** [comment]
this was a fast contribution :), in general looks pretty good, I've added some comments for improvements.

File: plugins/out_datadog/datadog.c [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    /* Get upstream connection */
+    upstream_conn = flb_upstream_conn_get(ctx->upstream);
+    if (!upstream_conn) {
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+
+    /* Convert input data into a Datadog JSON payload */
+    ret = datadog_format(data, bytes, tag, tag_len, &payload_buf, &payload_size, ctx);
+    if (ret == -1) {
+        FLB_OUTPUT_RETURN(FLB_ERROR);
```
[the remark this answers is not in the document]
Comment (clamoriniere) [reply plugins/out_datadog/datadog.c]: good catch !

File: plugins/out_datadog/datadog.c (hunk 1 above)
Comment (clamoriniere) [reply plugins/out_datadog/datadog.c]: Yes, saw it after fixing your first comment :)

File: plugins/out_datadog/datadog.c (hunk 3 above)
Comment (clamoriniere) [reply plugins/out_datadog/datadog.c]: ouch that was bad... (wrong copy/past)

**Comment by clamoriniere** [comment]
Hi @edsiper
This was also a fast review. I updated the PullRequest according to yours comments.

Appveyor build is broken but it doesn't seams related to my changes. (other pull-requests build are in error too)

**Comment by edsiper** [comment]
doing a simple run it crashes, please check unitialized variables and validate if the context is set or not, e.g:

```
$ bin/fluent-bit -i random -p samples=1 -o datadog -f 1
Fluent Bit v1.3.0
Copyright (C) Treasure Data

[2019/08/16 16:50:47] [ info] [storage] initializing...
[2019/08/16 16:50:47] [ info] [storage] in-memory
[2019/08/16 16:50:47] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/08/16 16:50:47] [ info] [engine] started (pid=26957)
[engine] caught signal ([2019/08/16 16:50:47] [error] [out_datadog] no ApiKey configuration key defined
SIGSEGV)
#0  0x561d8465f555      in  flb_upstream_destroy() at src/flb_upstream.c:132
#1  0x561d846ad08e      in  flb_datadog_conf_destroy() at plugins/out_datadog/datadog_conf.c:234
#2  0x561d846aca1b      in  flb_datadog_conf_create() at plugins/out_datadog/datadog_conf.c:97
#3  0x561d846ab820      in  cb_datadog_init() at plugins/out_datadog/datadog.c:39
#4  0x561d84653be8      in  flb_output_init() at src/flb_output.c:561
#5  0x561d8465a362      in  flb_engine_start() at src/flb_engine.c:412
#6  0x561d8464317f      in  main() at src/fluent-bit.c:853
#7  0x7f4d644e8b6a      in  ???() at ???:0
#8  0x561d84641369      in  ???() at ???:0
#9  0xffffffffffffffff  in  ???() at ???:0
Aborted
```

I suggest you try the tests under valgrind

**Comment by clamoriniere** [comment]
Hi @edsiper ,
Sorry about that.
I fixed the crash end tested severals case with valgrind. It should be fine now.

**Comment by irabinovitch** [comment]
Thanks @clamoriniere   

@edsiper appreciate the reviews!

**Comment by edsiper** [comment]
I signed up with a free trial and trying to ingest my CPU metrics using the given API key, but I get this response from the server:

```
[2019/08/23 14:26:56] [error] [out_datadog] http-intake.logs.datadoghq.com:0, HTTP status=403
```

maybe I am using the wrong API key ?, how to set it up?

**Comment by clamoriniere** [comment]
Hi @edsiper 

I updated a bit the plugin for improving the  `no tls` support it should be better now.

```console
$ bin/fluent-bit -i cpu  -o datadog  -p apiKey=<api-key>
Fluent Bit v1.3.0
Copyright (C) Treasure Data

[2019/08/25 19:36:36] [ info] [storage] initializing...
[2019/08/25 19:36:36] [ info] [storage] in-memory
[2019/08/25 19:36:36] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/08/25 19:36:36] [ info] [engine] started (pid=10917)
[2019/08/25 19:36:36] [ info] [sp] stream processor started
[2019/08/25 19:36:41] [ info] [out_datadog] http://http-intake.logs.datadoghq.com, port=80, HTTP status=200 payload={}
[2019/08/25 19:36:46] [ info] [out_datadog] http://http-intake.logs.datadoghq.com, port=80, HTTP status=200 payload={}
[2019/08/25 19:36:51] [ info] [out_datadog] http://http-intake.logs.datadoghq.com, port=80, HTTP status=200 payload={}
```

for `tls` connection support
```console
$ bin/fluent-bit -i cpu  -o datadog  -p apiKey=<api-key> -p tls=on 
Fluent Bit v1.3.0
Copyright (C) Treasure Data

[2019/08/25 19:45:16] [ info] [storage] initializing...
[2019/08/25 19:45:16] [ info] [storage] in-memory
[2019/08/25 19:45:16] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/08/25 19:45:16] [ info] [engine] started (pid=10925)
[2019/08/25 19:45:16] [ info] [sp] stream processor started
[2019/08/25 19:45:21] [ info] [out_datadog] https://http-intake.logs.datadoghq.com, port=443, HTTP status=200 payload={}
```

In case of invalid `apikey` you shoud see
```console
$ bin/fluent-bit -i cpu  -o datadog  -p apiKey=<wrong-api-key>
Fluent Bit v1.3.0
Copyright (C) Treasure Data

[2019/08/25 19:44:56] [ info] [storage] initializing...
[2019/08/25 19:44:56] [ info] [storage] in-memory
[2019/08/25 19:44:56] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/08/25 19:44:56] [ info] [engine] started (pid=10923)
[2019/08/25 19:44:56] [ info] [sp] stream processor started
[2019/08/25 19:45:01] [error] [out_datadog] https://http-intake.logs.datadoghq.com:443 HTTP status=403
[2019/08/25 19:45:06] [error] [out_datadog] https://http-intake.logs.datadoghq.com:443 HTTP status=403
```

or if the `apikey` don't have the correct length, you should receive a `status=400`

Let me know if you still have an issue with the plugin validation

**Comment by clamoriniere** [comment]
Hi @edsiper 
We support both, HTTP and HTTPS but we the recommended solution is HTTPS.
I didn't see in the plugin developer documentation a way to set by default `TLS=on`. Is it possible?

During the crash investigation, I improved the way we configure the plugin. It should have fix the memory issue.

Thanks again for your support on this PR. 👍

**Comment by edsiper** [comment]
thanks for the fixes.

I am a bit stuck in testing, I tried API Key and App API key, please drop me an email so we can have a conf to speed up the test.

**Comment by edsiper** [comment]
Tests were done. Everything looks good.

**Comment by clamoriniere** [comment]
@edsiper Thanks a lot for your support !

**Pull request merged**